federation/pkg/federation-controller/sync/controller.go

```diff
@@ -453,33 +453,70 @@ func syncToClusters(clustersAccessor clustersAccessorFunc, operationsAccessor op
 type clusterObjectAccessorFunc func(clusterName string) (interface{}, bool, error)
 
 // clusterOperations returns the list of operations needed to synchronize the state of the given object to the provided clusters
-func clusterOperations(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, obj pkgruntime.Object, accessor clusterObjectAccessorFunc) ([]util.FederatedOperation, error) {
-	key := federatedtypes.ObjectKey(adapter, obj)
-	operations := make([]util.FederatedOperation, 0)
+func clusterOperations(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, fedObj pkgruntime.Object, accessor clusterObjectAccessorFunc) ([]util.FederatedOperation, error) {
+	key := federatedtypes.ObjectKey(adapter, fedObj)
+	currentObjs := make(map[string]pkgruntime.Object)
 	for _, cluster := range clusters {
 		clusterObj, found, err := accessor(cluster.Name)
 		if err != nil {
 			wrappedErr := fmt.Errorf("Failed to get %s %q from cluster %q: %v", adapter.Kind(), key, cluster.Name, err)
 			runtime.HandleError(wrappedErr)
 			return nil, wrappedErr
 		}
-		// The data should not be modified.
-		desiredObj := adapter.Copy(obj)
 
-		var operationType util.FederatedOperationType = ""
 		if found {
-			clusterObj := clusterObj.(pkgruntime.Object)
-			if !adapter.Equivalent(desiredObj, clusterObj) {
+			currentObjs[cluster.Name] = clusterObj.(pkgruntime.Object)
+		} else {
+			currentObjs[cluster.Name] = nil
+		}
+	}
+
+	scheduledObjs := make(map[string]pkgruntime.Object)
```

Comment:
I think determination of the scheduled objects is a good candidate for extraction into a separate function.  Input: adapter, fed obj and cluster objs.   Output: scheduled objs.  This will simplify both maintenance and testing.